pycbc/events/coinc_rate.py

```diff
@@ -0,0 +1,156 @@
+#
+# =============================================================================
+#
+#                                   Preamble
+#
+# =============================================================================
+#
+""" This modules contains functions for calculating and manipulating
+coincident triggers.
+"""
+
+import itertools
+import numpy
+import pycbc.detector
+
+
+def multiifo_noise_coinc_rate(rates, ifos, slop):
+    """
+    Calculate the expected rate of noise coincidences for multiple detectors
+
+    Parameters
+    ----------
+    rates: list of lists/arrays
+        List of vector of single-detector trigger rates
+        in order of template id
+    ifos: list of strings
+        list of ifos in e.g. 'H1' form
+    slop: float
+        time to be added to time-of-flights between detectors to account
+        for timing error
+
+    Returns
+    -------
+    expected_coinc_rates: dict
+        A dictionary of ifo combined rates keyed on the combination string
+        value is the expected coincidence rate in the combination given the
+        individual detector trigger rates
+    """
+    ifos = numpy.array(ifos)
+    rates = numpy.array(rates)
+    expected_coinc_rates = {}
+    n_ifos = len(ifos)
+    assert len(rates) == n_ifos
+    # Calculate coincidence for all-ifo combination
+    # multiply the two rates and by the overlap time
+    allowed_area = multiifo_noise_coincident_area(ifos, slop)
+    rateprod = [numpy.prod(rs) for rs in zip(*rates)]
+    ifostring = ''.join(ifos)
+    expected_coinc_rates[ifostring] = allowed_area * numpy.array(rateprod)
+    # if more than one possible coicidence type exists,
+    # calculate coincidences for subsets
+    if n_ifos > 2:
+        # Calculate rate for each 'miss-one-out' detector combination
+        subsets = itertools.combinations(ifos, n_ifos-1)
+        for subset in subsets:
+            i_set = [numpy.nonzero(ifo == ifos)[0][0] for ifo in subset]
+            # calculate coincidence rates in subsets through iteration
+            sub_coinc_rates = multiifo_noise_coinc_rate(rates[i_set],
+                                                        ifos[i_set], slop)
+            # add these sub-coincidences to the overall dictionary
+            for sub_coinc in sub_coinc_rates:
+                expected_coinc_rates[sub_coinc] = sub_coinc_rates[sub_coinc]
+
+    return expected_coinc_rates
+
+
+def multiifo_noise_coincident_area(ifos, slop):
+    """
+    calculate the multiplicative factor of the individual detector trigger
```

Comment:
prefer to say what calculation the function is doing. eg 'total extent of time offset between 2 detectors, or area of the 2d space of time offsets for 3 detectors, for which a coinc can be generated'